### PR TITLE
koordlet: support core sched cookie management

### PR DESCRIPTION
**Pull request by saintube**
### Ⅰ. Describe what this PR does

koordlet: support core scheduling cookie management of containers.

1. apis: Add pod-level API #1720.
2. koordlet/utils: Add core sched basic operations via prctl (Get/Create/ShareTo/ShareFrom). Add enhanced operations like Assign/Clear (LockOSThread required).
3. koordlet/runtimehooks: Add the plugin CoreSched to manage containers' cookies.

### Ⅱ. Does this pull request fix one issue?

Part of #1728.

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/runtimehooks/hooks/coresched/rule.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	_, ok := e.(*struct{})
+	if !ok {
+		return false, fmt.Errorf("invalid rule type %T", e)
+	}
+
+	if p.allPodsSyncOnce.Load() {
+		return false, nil
+	}
+
+	p.allPodsSyncOnce.Store(true)
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/coresched/rule.go:132]: where is the ref of `allPodsSyncOnce`

Comment (saintube) [reply pkg/koordlet/runtimehooks/hooks/coresched/rule.go:132]: L128

File: apis/slo/v1alpha1/pod.go [hunk 2]
```diff
@@ -64,3 +63,35 @@ func GetPodMemoryQoSConfig(pod *corev1.Pod) (*PodMemoryQOSConfig, error) {
...
+	// When the Core Sched is enabled, pods with the different core sched group IDs will not be running at the same SMT
+	// core at the same time, which means they will take different core sched cookies. If a pod sets the core sched
+	// disabled, it will take the default core sched cookie (0) and will also be force-idled to run on the same SMT core
+	// concurrently with the core-sched-enabled pods.
+	// Value should be a valid UUID or the none value "0".
+	// When the value is a valid UUID, pods of that group ID on the node will be assigned to the same core sched cookie.
+	// When the value is the none value "0", pod will be reset to the default core sched cookie `0`.
+	// When the annotation is missing but the node-level strategy enables the core sched, the pod will be assigned an
+	// internal group according to the pod's UID.
+	AnnotationCoreSchedGroupID = apiext.DomainPrefix + "core-sched-group-id"
```
[the remark this answers is not in the document]
Comment (saintube) [reply apis/slo/v1alpha1/pod.go]: done

File: apis/slo/v1alpha1/nodeslo_types.go [hunk 3]
```diff
@@ -25,10 +25,21 @@ import (
 
 // CPUQOS enables cpu qos features.
 type CPUQOS struct {
+	// applied policy for the CPU QoS, default = "groupIdentity"
+	Policy *CPUQOSPolicy `json:"policy,omitempty"`
 	// group identity value for pods, default = 0
 	GroupIdentity *int64 `json:"groupIdentity,omitempty" validate:"omitempty,min=-1,max=2"`
```
[the remark this answers is not in the document]
Comment (saintube) [reply apis/slo/v1alpha1/nodeslo_types.go]: added

File: pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+//     The cookies of sibling PGIDs (i.e. the PGIDs of the same core sched group) will be fetched in the Assign. If all
+//     the cookies of sibling PGIDs are default or invalid, the Assign will fall back to Add.
+//  4. Clear: Clear a cookie of an existing core sched group for a container (reset to default cookie 0), and the
+//     cookie's reference count in the cache is decreased. The cache entry will be removed when the ref count is no
+//     larger than zero.
+//
+// If multiple non-default cookies are assigned to existing containers of the same group, the firstly-created and
+// available cookie will be retained and the PGIDs of others will be moved to the former.
+// NOTE: The agent itself should be set the default cookie. It can be excluded by setting QoS to SYSTEM.
+func (p *Plugin) SetContainerCookie(proto protocol.HooksProtocol) error {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go:188]: - add rateLimiter for safety
- tradeoff between /proc read and syscall

File: pkg/koordlet/statesinformer/impl/states_informer.go [hunk 5]
```diff
@@ -170,6 +169,9 @@ func (s *statesInformer) Run(stopCh <-chan struct{}) error {
 		}
 	}
 
+	// start callback runner after informers synced
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/statesinformer/impl/states_informer.go:172]: done

File: pkg/koordlet/config/config_test.go [hunk 6]
```diff
@@ -30,6 +30,7 @@ func TestConfiguration_InitFlags(t *testing.T) {
 		"--reconcile-interval-seconds=5",
 		"--cgroup-root-dir=/host-cgroup/",
 		"--feature-gates=AllBeta=true,AllAlpha=false",
+		"--enable-kernel-core-sched=false",
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/config/config_test.go]: removed

File: pkg/koordlet/resourceexecutor/cgroup.go [hunk 7]
```diff
@@ -193,6 +193,31 @@ func readCgroupAndParseInt32Slice(parentDir string, r sysutil.Resource) ([]int32
...
+// readCgroupAndParseUint32Slice reads the given cgroup content and parses it into an uint32 slice.
+// e.g. content: "1\n23\n0\n4\n56789" -> []int32{ 1, 23, 0, 4, 56789 }
+func readCgroupAndParseUint32Slice(parentDir string, r sysutil.Resource) ([]uint32, error) {
+	s, err := cgroupFileRead(parentDir, r)
+	if err != nil {
+		return nil, err
+	}
+
+	// content: "%d\n%d\n%d\n..."
+	var values []uint32
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/resourceexecutor/cgroup.go:205]: it is a general helper function, perhaps we can refactor to golang generics later

File: pkg/koordlet/util/system/core_sched.go [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+	}
+	if len(failedPIDs) > 0 {
+		return failedPIDs, fmt.Errorf("assign cookie for pidsTo error")
+	}
+	return nil, nil
+}
+
+// IsCoreSchedSupported checks if the core scheduling feature is enabled in the kernel sched_features.
+// If `--enable-kernel-core-sched=true`, it tries to enables the core scheduling feature in the kernel.
+func IsCoreSchedSupported() (bool, string) {
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/util/system/core_sched.go]: done

File: pkg/koordlet/util/system/core_sched.go [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+	SchedFeatureNoCoreSched = "NO_CORE_SCHED"
+)
+
+// CoreSchedScopeType defines the type of the PID type operated in core sched.
+type CoreSchedScopeType uint
+
+const (
+	// CoreSchedScopeThread means the PID type operated in core sched is a thread.
+	CoreSchedScopeThread CoreSchedScopeType = iota
+	// CoreSchedScopeThreadGroup means the PID type operated in core sched is a thread group.
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/util/system/core_sched.go:40]: switch to thread and thread group mode

File: pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go (hunk 4 above)
Comment (saintube) [reply pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go:189]: no longer read /proc, switch to prctl for thread groups

File: pkg/koordlet/resourceexecutor/cgroup.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (zwzhang0107) [reply pkg/koordlet/resourceexecutor/cgroup.go:205]: add a todo both here and ParseCgroupProcs

File: pkg/koordlet/util/system/core_sched.go [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+	if len(failedPIDs) > 0 {
+		return failedPIDs, fmt.Errorf("assign cookie for pidsTo error")
+	}
+	return nil, nil
+}
+
+// EnableCoreSchedIfSupported checks if the core scheduling feature is enabled in the kernel sched_features.
+// If kernel supported, it tries to enable the core scheduling feature in the kernel.
+// The core sched's kernel feature is known set in two places, if both of them are not found, the system is considered
+// unsupported for the core scheduling:
```
Comment (zwzhang0107) [inline pkg/koordlet/util/system/core_sched.go:259]: 1. sched_core
2. sched_features for old versions

File: pkg/koordlet/metrics/metrics.go [hunk 11]
```diff
@@ -32,6 +32,7 @@ func init() {
 	prometheus.MustRegister(CPUSuppressCollector...)
 	prometheus.MustRegister(CPUBurstCollector...)
 	prometheus.MustRegister(PredictionCollectors...)
+	prometheus.MustRegister(CoreSchedCollector...)
```
[the remark this answers is not in the document]
Comment (zwzhang0107) [reply pkg/koordlet/metrics/metrics.go:35]: leave a new issue?

File: pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go [hunk 12]
```diff
@@ -0,0 +1,478 @@
...
+	ruleNameForAllPods = name + " (allPods)"
+
+	defaultCacheExpiration     = 300 * time.Second
+	defaultCacheDeleteInterval = 600 * time.Second
+
+	// ExpellerGroupSuffix is the default suffix of the expeller core sched group.
+	ExpellerGroupSuffix = "-expeller"
+)
+
+// SYSTEM QoS is excluded
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go:56]: clarify the policy for SYSTEM QoS, use one cookie?

File: pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+	isEnabled, groupID := p.getPodEnabledAndGroup(podAnnotations, podLabels, podKubeQOS, podUID)
+	klog.V(6).Infof("manage cookie for container %s/%s, isEnabled %v, groupID %s",
+		podMetaName, containerName, isEnabled, groupID)
+
+	lastGroupID, lastCookieEntry, cookieEntry := p.getCookieCacheForContainer(groupID, containerUID)
+
+	// expect enabled
+	// 1. disabled -> enabled: Add or Assign.
+	// 2. keep enabled: Check the differences of cookie, group ID and the PIDs, and do Assign.
+	if isEnabled {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go:227]: this is a BIG if/else, could it be splited?

File: pkg/koordlet/metrics/metrics.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/metrics/metrics.go:35]: https://github.com/koordinator-sh/koordinator/issues/1728

File: pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go (hunk 13 above)
Comment (saintube) [reply pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go:221]: done

**Comment by zwzhang0107** [comment]
make sure that system qos will not be hungry to death (share value too small)

**Review by jasonliu747 — APPROVED** [review]
/lgtm

File: pkg/koordlet/runtimehooks/hooks/coresched/cookie_cache.go [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+type IndexedValue[T constraints.Ordered] struct {
+	Value T
+	Index int
+}
+
+const MaxSimpleRecreateSize = 4
+
+// OrderedMap is an ordered map with type T.
+type OrderedMap[T constraints.Ordered] struct {
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/runtimehooks/hooks/coresched/cookie_cache.go]: Changed to a simple map structure.
Actually, the `assignCookie` method in helper.go needs an ordered pids to shareFrom the earliest sibling pid and reduce randomness.

File: pkg/koordlet/util/system/core_sched.go [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+	isSysctlSupported, err := GetSchedCore()
+	if err == nil && isSysctlSupported {
+		klog.V(6).Infof("Core Sched is already enabled by sysctl")
+		return true, ""
+	}
+	if err == nil { // sysctl supported while value=0
+		klog.V(6).Infof("Core Sched is disabled by sysctl, try to enable it")
+		err = SetSchedCore(true)
+		if err == nil {
+			klog.V(4).Infof("Core Sched is enabled by sysctl successfully")
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/util/system/core_sched.go]: fixed

File: pkg/koordlet/util/system/core_sched.go [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+	CoreSchedScopeThread CoreSchedScopeType = iota
+	// CoreSchedScopeThreadGroup means the PID type operated in core sched is a thread group.
+	CoreSchedScopeThreadGroup
+	// CoreSchedScopeProcessGroup means the PID type operated in core sched is a process group.
+	CoreSchedScopeProcessGroup
+)
+
+// CoreSchedInterface defines the basic operations of the Linux Core Scheduling.
+// https://docs.kernel.org/admin-guide/hw-vuln/core-scheduling.html
+type CoreSchedInterface interface {
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/koordlet/util/system/core_sched.go:48]: comments added

File: pkg/koordlet/runtimehooks/hooks/coresched/cookie_cache.go [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+func (p PIDCache) Len() int {
+	return len(p)
+}
+
+func (p PIDCache) Has(pid uint32) bool {
+	_, ok := p[pid]
+	return ok
+}
+
+func (p PIDCache) GetAll() []uint32 {
```
Comment (hormes) [inline pkg/koordlet/runtimehooks/hooks/coresched/cookie_cache.go:137]: GetAll -> GetAllSorted

Comment (saintube) [reply pkg/koordlet/runtimehooks/hooks/coresched/cookie_cache.go]: fixed

**Review by hormes — APPROVED** [review]

**Pull request merged**